### PR TITLE
contracts: Add ROFL subcall rofl.IsAuthorizedOrigin

### PR DESCRIPTION
**Pull request by kostko**

### CODE REVIEW

File: contracts/contracts/Subcall.sol [hunk 1]
```diff
@@ -578,4 +586,20 @@ library Subcall {
...
+
+    /**
+     * @notice Verify whether the origin transaction is signed by an authorized
+     * ROFL instance for the given application.
+     * @param appId ROFL app identifier
+     */
+    function roflEnsureAuthorizedOrigin(bytes21 appId) internal {
+        (uint64 status, bytes memory data) = subcall(
+            ROFL_IS_AUTHORIZED_ORIGIN,
+            abi.encodePacked(hex"55", appId)
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:598]: Is `hex"55"` some CBOR magic constant? Maybe add a comment for this similar to `consensusDelegate()` above.

Comment (peternose) [inline contracts/contracts/Subcall.sol:598]:
```suggestion
            abi.encodePacked(hex"55", appId) // byte string, 21 bytes
```

**Review by matevz — APPROVED** [review]
Please comment specific magic numbers. Otherwise LGTM.

File: contracts/contracts/Subcall.sol [hunk 2]
```diff
@@ -578,4 +586,20 @@ library Subcall {
...
+        );
+
+        if (status != 0 || data[0] != 0xf5) {
```
Comment (peternose) [inline contracts/contracts/Subcall.sol:601]: Do we need to check if `data.len == 1`?

File: contracts/contracts/Subcall.sol (hunk 1 above)
Comment (kostko) [reply contracts/contracts/Subcall.sol:598]: It stands for "CBOR byte-array of length 21 bytes".

File: contracts/contracts/Subcall.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (kostko) [reply contracts/contracts/Subcall.sol:601]: CBOR encoded "boolean true".

**Comment by kostko** [comment]
Added comments, please take another look.

File: contracts/contracts/Subcall.sol [hunk 3]
```diff
@@ -578,4 +586,22 @@ library Subcall {
...
+     */
+    function roflEnsureAuthorizedOrigin(bytes21 appId) internal {
+        (uint64 status, bytes memory data) = subcall(
+            ROFL_IS_AUTHORIZED_ORIGIN,
+            abi.encodePacked(hex"55", appId) // CBOR byte string, 21 bytes.
+        );
+
+        // The result should be a CBOR-encoded boolean with the value true indicating
+        // that the origin is authorized for the given ROFL app.
+        if (status != 0 || data.len != 1 || data[0] != 0xf5) {
```
Comment (peternose) [inline contracts/contracts/Subcall.sol:603]:
```suggestion
        if (status != 0 || data.length != 1 || data[0] != 0xf5) {
```

**Review by peternose — APPROVED** [review]

**Comment by kostko** [comment]
Also added some basic tests.

**Comment by aefhm** [comment]
> Oh, contracts tests are not required for merging 🤦

Yeah, I was wondering about the test setup.

**Comment by matevz** [comment]
> I fixed this after making the comment, so they should be required now ;)

I [tested this yesterday](https://github.com/oasisprotocol/sapphire-paratime/pull/354) and can confirm it.

**Comment by aefhm** [comment]
> > I fixed this after making the comment, so they should be required now ;)
> 
> I [tested this yesterday](https://github.com/oasisprotocol/sapphire-paratime/pull/354) and can confirm it.

Oh sorry. I meant a new test for the ROFL subcall.

**Pull request merged**